### ISSUE TITLE
Minor privacy leak between two parties on the same participant

### ISSUE THREAD

**Issue by bame-da** [digital-asset/daml#7561]
Archived contracts are immediately removed from the Index Database, which makes it impossible to perform authorization checks on the Fetch node thereafter. As a result, an unauthorized party that has once had a contract divulged to them, and which is colocated with a stakeholder of that contract can find out whether that contract sill exists.

Deploy these templates:

```
template T
  with
    sig : Party
  where
    signatory sig

template TDivulger
  with
    sig : Party
    obs : Party
  where
    signatory sig
    observer obs

    choice Fetch : T
      with
        t : ContractId T
        ct : Party
      controller ct
      do fetch t
```

Then in DAML Repl:

```
daml> sig <- allocateParty "sig"
daml> obs <- allocateParty "obs"
daml> tcid <- submit sig do createCmd T with sig
daml> _ <- submit sig do createAndExerciseCmd (TDivulger with ..) (Fetch tcid sig)

daml> submit obs do createAndExerciseCmd (TDivulger with sig = obs; obs) (Fetch tcid obs)
... Error: node NodeId(2) requires one of the stakeholders TreeSet(party-d40bb148) of the fetched contract to be an authorizer...

daml> submit sig do exerciseCmd tcid Archive

daml> submit obs do createAndExerciseCmd (TDivulger with sig = obs; obs) (Fetch tcid obs)
...couldn't find contract ContractId(...
```

In other words, `obs` learns whether `tcid` still exists by virtue of either getting an authorization error, or a contract not found error.

**Comment by bame-da** [comment]
My feeling is that a possible resolution would be to go the direction of @hurryabit's idea to split out "multi-party submission" read and write from each other. If I understood correctly the idea would be to use DAML's authorization context to authorize fetches and key lookups rather than the requesters authority. 

The status quo is that when a `fetch` or `fetchByKey` is encountered, the engine calls the index database to look up the key or contractId. That call includes the requester, and the lookup only succeeds if the requester has had the contract divulged or is a stakeholder, respectively. 

We could modify that to instead use the authorisers of the `fetch` or `fetchByKey` according to the definitions in the [Ledger Model](https://docs.daml.com/concepts/ledger-model/ledger-integrity.html#da-model-authorization). In case of `fetch`, the lookup would succeed exactly if the contract is available in the index, and of the of the authorizers is a stakeholder of the contract. A `fetchByKey` would succeed if one of the authorizers is a stakeholder of the contract _and is hosted on the participant_. The latter rule ensures we don't run into cases where there are multiple contracts for a given key. `lookupByKey` should change it's behaviour in line with `fetchByKey` in the sense that it should resolve to `Some cid` exactly if `fetchByKey` would succeed. The authorization rules in DAML would remain unchanged.

@hurryabit @meiersi-da @gerolf-da @cocreature , what do you think?

**Comment by cocreature** [comment]
Impilcitly relying on party topology seems very confusing to me. If I understand your proposal correctly this is also not backwards compatible whereas the current proposal for multi-party submissions is.

**Comment by bame-da** [comment]
Why do you say it's not backwards compatible? Any successful command submission now would also be a successful command submission with this change.

Some error messages would change, but since they currently cause a privacy leak, that's a bug-/security fix.
Some transactions that would error now would succeed with this change. That's a feature.

**Comment by cocreature** [comment]
As far as I understand, a submission that fails on the submitter check could succeed under your proposal. I would consider that a breaking change. Maybe I deliberately designed my model this way.

**Comment by bame-da** [comment]
That is true, this may have to be an opt-in feature, just like I'm suggesting it would have to be on Script and Sandbox.

**Comment by meiersi-da** [comment]
@bame-da AFAIR, you are working on a series of proposals that removes the need for divulgence

**Comment by meiersi-da** [comment]
@bame-da what about the option to fix this in production by mapping all authorization errors in fetches to `contract not found`.

**Comment by bame-da** [comment]
@meiersi-da the problem actually extends to exercises as well so we'd have to reduce all authorization failures to simple "fetch/exercise failed" messages without any further information. It's a possible solution, but the UX impact is quite heavy in my opinion. A middle ground may be to log the actual error, and only return a "fetch/exercise failed" with an ID via the API. In that case admins could get at the actual problem, at least.

**Comment by bame-da** [comment]
Fundamentally I believe that divulged contracts should not be available for interpretation at all as they lead exactly to the problem that you can never remove them from the Index DB or your client application's operational data store. But that's a deep change, and we don't have a better disclosure mechanism in place yet so not an option as the immediate solution.

What if we retained contracts conditionally on who previously witnessed it? The Index already has the information which hosted parties know of a contract. It needs that to do the authorization check on fetches. When it receives a consuming exercise for the contract, it could also get the information which (hosted) parties witness that exercise. Could we not gate the removal of the contract on every party that knows of the contract also witnessing the archive? 
If we did that, we would get equal behaviour whether two parties are colocated or not.

**Comment by meiersi-da** [comment]
@bame-da I agree with divulgence needing to go, and I'm hesitant to investing further into it. I'd much rather fix it.

So for this problem, I'd favor to make redacting authorization errors a configuration. This way production ledgers can turn it on, while for development it is turned off. We can then schedule this small investment into divulgence vs. investments into replacing divulgence.

**Comment by bame-da** [comment]
From my point of view my proposal above (retaining contracts in the index conditionally) has the advantage of improved UX and consistency between topologies. What's the advantage of making the error messages configurable?
